### DISCUSSION TITLE
State of romcal and its latest rewrite

### DISCUSSION THREAD

**Discussion by tukusejssirs** [romcal/romcal#445] [General]
Hey, @emagnier!

I’d like to know what’s your status regarding romcal. Last time you told me you are rewriting romcal in a monorepo, but you didn’t want to share with me any code (unfortunately) until you do some more stuff on it.
As I see it, you are busy with other stuff (family, work, other projects and hobbies), therefore sharing the dev burden and making the dev process open from the very beginning is IMHO the way to go.

Do you agree to share with me the details about the rewrite? Like:

- What are the issues you to solve with the rewrite?
- Where can I access the rewrite code?
- What did you do?
- What do you want to do still?
- Is it still MIT-licensed?
- Are there any other details you want to share with me?

---

I hope you’re doing well! :wink:

**Comment by emagnier** [comment]
Hello @tukusejssirs 👋 

Thanks for asking about the project's status. 
I'm trying to answer your questions below (and not forget anything).

### What are the issues you to solve with the rewrite?

1. Significant testing gap.
2. The lack of codebase modularization complicates testing, and restricts the potential for contribution, evolution, or extension of the Romcal library.
3. Data contained in TypeScript files, rather than being language-agnostic.
4. Some limitations in the current calendar inheritance functionality (such as handling countries across multiple continents).
5. Build scripts and NPM publishing is cumbersome and overly complex (to use & to maintain).

### Where can I access the rewrite code?

This work is accessible from this branch, which I pushed some time ago (though I haven't announced it here yet): 
https://github.com/romcal/romcal/tree/optimizations-ex.

### What did you do?

1. Codebase refactoring and modularization: to enhance the ability to conduct unit and integration testing, ease the evolution of the codebase, and simplify the extension or replacement of certain built-in features of Romcal.
2. Language-agnostic data: make the data independent of the programming language used in the library. This allows the data to be more easily utilized by libraries other than Romcal, and conversely, enable Romcal to use alternative data sources.
3. Streamline future data contributions: use a human-friendly data format (YAML). Provide some ready to use configurations for VSCode (one of the most used IDEs, and now built-in with GitHub), to enable autocompletion (intellisense) and data validation while contributing from VSCode.
4. Strengthen typing, data quality, and documentation: enhance data typing, improve data quality, and provide comprehensive documentation. Generate JSON-Schema documentation to ensure clear understanding of the data structure (for data usage from other libraries or programming languages).
5. Revamp the calendar inheritance feature: redesign the calendar inheritance system to manage complexities such as handling countries across multiple continents.
6. Enable the addition of more calendars, such as diocesan or parish calendars, through a reorganization of calendar tree-files (folder structure and calendar naming to be discussed), thanks to the redesigned calendar inheritance system.
7. Simplify and streamline build process.

### What do you want to do still?

1. The refactor of the packages/core is not yet complete, and there are still many tests to add, and some fine-tuning to do here and there.
2. Adopt NX to manage scripts across all the packages in the monorepo.
3. About data, I also want to add data validation scripts, to identify gaps in the data, and establish a clear contribution process.
4. Rework the publishing process on NPM. Update the CI tools.
5. Undertake a comprehensive documentation effort: update `.md` files and utilize Docusaurus to package and ease access to this documentation.

### Is it still MIT-licensed?

Yes. Unless you have any points to share here?

### Are there any other details you want to share with me?

1. All the work done in this branch, as well as the directions taken, are not set in stone and are open for discussion.
2. I initially undertook this work to address personal project needs, and to do some experiments. I got caught up in the whirlwind, and it's turned out to be a lot more work than anticipated. Due to time constraints and the need to make rapid progress, I adopted the approach I used in this branch (which is still unfinished due to personal time limitations). Looking back, I strongly believe that the direction taken in this branch will simplify the codebase, improve the developer experience, and consequently, facilitate future contributions to this open-source project. This is why I'm excited to share this work here!
3. To make further progress, I require time. Currently, I have very little time available, but I hope to dedicate more time to it in the near future.
4. Lastly, I don't intend for this `optimizations-ex` branch to be a roadblock for anything. Currently, the main development branch is `dev`, and it should remain that way. I will ensure to keep the `optimizations-ex` branch updated (from the `dev` branch) as contributions are made to the `dev` branch.

**Reply by tukusejssirs** [reply]
Hey, @emagnier!

Sorry for being AWOL, but work and family duties caught me up longer than expected.

I haven’t been able to check the new code yet, but the progress is awesome! :100: :tada:

I have some additional questions:

1. Did you work on the start/end times and celebration/l10n sources too?
2. Nice to hear you want to use Nx monorepo. What made you change your opinion? As far as I remember, you wanted to use a different monorepo utility (Turborepo?).

Notes:

> Is it still MIT-licensed?
>
> Yes. Unless you have any points to share here?

Well, I just wanted to be sure whether you still want to keep it open-source, as you started building it behind the scenes. :wink:

> To make further progress, I require time. Currently, I have very little time available, but I hope to dedicate more time to it in the near future.

I definitely want to contribute to the code, although time constraints are on my side too. Hopefully, I can find a few hours every week for romcal.

> Lastly, I don't intend for this `optimizations-ex` branch to be a roadblock for anything. Currently, the main development branch is `dev`, and it should remain that way. I will ensure to keep the `optimizations-ex` branch updated (from the `dev` branch) as contributions are made to the `dev` branch.

While I am okay with keeping the current `3.0.0-dev.*` development, I think once you think we are ready to merge `optimizations-ex` into `dev` and continue the development there. IMHO we should strive to have only one dev romcal version from which we should cut a stable release every now and then. WDYT? :thinking:

**Reply by emagnier** [reply]
Hey @tukusejssirs, great to hear from you again!

> Did you work on the start/end times and celebration/l10n sources too?

I'm not entirely sure I understand your question. No, I only converted the data into YAML, but I didn't modify them. I also didn't add any functionality to consider including a celebration only from a specific date.

> Nx vs. Turborepo

Up until now, I didn't have much experience with these two tools. After taking another close look at them, I simply became more convinced to test and adopt Nx (and especially since you mentioned that you also have experience with this tool). Help to set and use correctly this tool will be appreciated!

Lastly, I do plan to merge the `optimizations-ex` branch as soon as it's fully functional (at least without any known bugs or regressions).

**Reply by tukusejssirs** [reply]
> > Did you work on the start/end times and celebration/l10n sources too?
> 
> I'm not entirely sure I understand your question. No, I only converted the data into YAML, but I didn't modify them. I also didn't add any functionality to consider including a celebration only from a specific date.

Basically, this feature is related to #23, however, also to the current (post Vatican II) calendar, as is inserted/updated/moved from time to time, therefore I suggest to add a start date (for now defaults to be 1 Jan 1970 when undefined, later it should be a required property) and end date to all celebrations. I don’t expect to have full support of these properties. Initially, we should only care about the end date: when defined, ignore it altogether. This way we could fill the data without breaking anything really.

> > Nx vs. Turborepo
> 
> Up until now, I didn't have much experience with these two tools. After taking another close look at them, I simply became more convinced to test and adopt Nx (and especially since you mentioned that you also have experience with this tool). Help to set and use correctly this tool will be appreciated!

I’ll help, however, we should discuss some configuration (probably in a PR or an issue).

> Lastly, I do plan to merge the `optimizations-ex` branch as soon as it's fully functional (at least without any known bugs or regressions).

:+1:

**Comment by tukusejssirs** [comment]
Hey, @emagnier! So finally, I’ve found some time to write down my thoughts and questions which I’d like to discuss with you regarding romcal’s Nx monorepo adoption.

---

First of all, I’d like to talk about romcal division package-wise.

I suggest creating the following packages:

- `@romcal/sources`:
  - it would contain all the martyrology/calendar/l10n sources (all data required for bibliographical reference to either a link a hard-copy publication) along with the reference (-s) to the original source (when available online;
  - ideally, all sources available online or to us should be copied to Google Drive and that link should be considered an archive);
  - no page numbers or chapter/section heading should be included;
  - why I suggest to create a separate package for sources?
    - when done properly, there will be too many sources;
    - some sources will be cited multiple times;
    - we should simply reference an imported object and provide the page range, optionally with a chapter/section heading;
- `@romcal/core`:
  - it should take `@romcal/martyrology`, `@romcal/locales` and `@romcal/grc1969/calendars` to build the requested calendar(s);
  - it should be calendar-agnostic, both revision-wise (`grc1962` vs `grc1969`) and rite-wise (Roman Catholic, Greek Catholic, Orthodox);
  - I understand that currently we support only GRC 1969 only, however, I still hope that in far future, we could extend the support to other calendar types (revision-wise and rite-wise);
- `@romcal/martyrology`:
  - it should contain all locale-agnostic celebration (meta-) data;
  - it would reference the sources available in `@romcal/sources` with page number (-s), possibly also with a chapter/section heading when required; 
- `@romcal/grc1969/calendars`:
  - it should contain all celebration data to build the calendars in a particular GRC revision, both the GRC and all particular calendars;
  - later we could also have `@romcal/grc1962/calendars` (cf #23) or `@romcal/grc1969-ordinariate-north-america/calendars` (cf #407)
  - it would reference the sources available in `@romcal/sources` with page number (-s), possibly also with a chapter/section heading when required; 
- `@romcal/locales`:
  - it should contain everything needed to locale the l10n and calendar data, incl the default formatting per locale;
  - it would reference the sources available in `@romcal/sources` with page number (-s), possibly also with a chapter/section heading when required; 

Below is a diagram how the data should flow from the data source to the package that uses that data.

```mermaid
graph TD;
  S("@romcal/sources") --> L("@romcal/locales");
  S --> M("@romcal/martyrology");
  M --> L;
  L --> R("@romcal/core");
  C("@romcal/*/calendars") --> R;
  M --> C;
  S --> C;
```

After writing this all down, it came to my mind that expect of `@romcal/core`, everything could a library, not an app. We should consider putting the other packages above to `libs/`.

---

Now, regarding other parts of the romcal Nx monorepo, some of them you could be aware of, I just wanted to write them all down not to forget about them:

- add a `README.md` file to each package;
- move some internal libraries (e.g. `shared`) from `packages/` to `libs/` when they are shared across multiple packages:
  - all that means is to create one or more libraries that would contain the utils and functions, models, types, etc used by multiple packages (two or more);
  - the libraries could be either:
    - buildable separately and publishable to a package registry (e.g. to NPM); or
    - be an internal library (used by romcal internally, merged with the packages at build time);
- consider moving `martyrology.yaml` from `calendars` package to a separate package:
  - as I have already suggested above, IMHO, the martyrology is not related to the calendars, therefore, it should be moved to a separate package;
  - recently, you said the it should contain only personal celebrations (actually personal metadata), however, before your recent comment, I was under impression that it should contain _all_ celebration metadata, including metadata of event celebrations:
    - take for example `epiphany_of_the_lord`: it is including in `martyrology.ts` both in the [`dev` branch](https://github.com/romcal/romcal/blob/017fd8531d4bf534b23d2c92252e65f2768aa4be/lib/catalog/martyrology.ts#L1145-L1147) and in [your `locale-checks` branch](https://github.com/romcal/romcal/blob/a6aa8750e7c8cb0d6158e114b9212463c8145162/lib/catalog/martyrology.ts#L1142-L1144), which according to your logic should not be part of `martyrology.ts`;
    - that said, _Martyrologium Romanum_ actually contains the event celebrations (e.g. the [Epiphany](https://gitlab.com/tukusejssirs/martyrology/-/raw/master/st_out/085_mr2004.tif?ref_type=heads)), therefore, I now suggest keeping all event celebrations in there and create missing ones;
    - as a side note, we might want to split the file into multiple, smaller ones:
      - one file could contain personal metadata, another metadata of events;
- `calendars` package:
  - do we need `central-america` (sub-) continental calendar?
    - we could if there is a patronage celebrated on that subcontinent;
  - consider creating `eurasia` (super-) continental calendar and make `europe` and `asia` be dependent on it;
  - consider creating `australia-and-ocenia` (super-) continental calendar and make `australia` and `oceania` be dependent on it;
  - create `australia` continental calendar;
  - consider creating `index.ts` barrel file to re-export all the calendars;
- `locales`:
  - add a barrel file to re-export all the locales;
- do we really need `build`, `eslint-config`, `jest-config` as separate packages?
  - as for `eslint` and `jest` config files, there are located in the monorepo root folder, and the overrides are located in the app/lib folders (they include the global/monorepo-level config);
  - as for `build`, it could be a separate package, I just have no idea how exactly you want to utilise it:
    - we might want to merge `build` into `@romcal/core`;
- do we really need a separate `proper-of-time` package?
  - I would presume it could be merged with the GRC calendar in `calendars`;
- consider using [`yaml`](https://www.npmjs.com/package/yaml) instead of `js-yaml` (which seems to be unmaintained);
- all packages should have the same version:
  - Nx enforces this, but we should enforce it at runtime too.

---

WDYT? :thinking:

**Reply by emagnier** [reply]
Hey @tukusejssirs, thank you very much for sharing your thoughts!

I'm a bit short on time to address all your points, but here's some initial feedback.

First, I noticed that we took different approaches to architecting Romcal. I used a functional codebase architecture, splitting the code into several packages based on functional scope, with shared components and configurations, and concentrating data in a specific package (@romcal/data). On the other hand, you propose the opposite: multiple packages per data type (calendars, locales, martyrology) but consolidating the Romcal codebase into a single package.

These two approaches can complement each other. In reality, the data is already organized as you presented but in various subfolders within the @romcal/data package.

I think the objective of my architecture is well described in my previous response, in the "What did you do?" section. In a few words, the goal of my architecture is to isolate each module and business layer of the Romcal library, making it easier to test each component, make them interchangeable and extendable by third-party components, and evolve them more independently. Thus, the aim is to use or extend Romcal's features and data more easily with third-party tools. So I've split consequently what is necessary for the proper functioning of the library, the data, and the validation and build tools into several packages.

> as for `build`, it could be a separate package, I just have no idea how exactly you want to utilise it

This package is used by almost all the other packages and contains common helpers for compiling the codebase or data of each package, generating the corresponding JSON schemas, and soon, it will enable us to publish properly each package independently on npmjs.

> Add barrel `index.ts` files to re-export all the locales & calendars

This is no longer necessary because the data is no longer in TS files but in YAML files, and it's imported differently by Romcal's build tools. If you want to access them, you should use the corresponding agnostics JSON files, which are built in the `@romcal/data/dist` directory.

> The martyrology is not related to the calendars

Correction: In fact, Romcal's primary purpose is not to be a database but a library for generating liturgical calendars. So the first-class citizen here is the calendar. From there, we derive the metadata of individuals/persons through the martyrology (which enriches the metadata of a celebration) and the locales (to translate all these metadata). The martyrology and locales are in service of the calendars, and not the other way around or independent entities.

Regarding support for other forms/rites, I've left several comments, including [this one](https://github.com/romcal/romcal/issues/23#issuecomment-1003461132). For now, I prefer to focus on modularizing the Romcal library, which will allow to easily "extend" Romcal to support such calendars but from a third-party repository/project. Therefore, I am against the direct support of other form/rites within Romcal because the context, rules and logics are often very different. It is essential to handle and test them separately.

As for storing sources or source references, that's a great idea. However, we need to consider copyright issues. For references, I believe they should be placed as close to the data as possible (closed to definitions of calendars/locales/martyrology...). Perhaps we should create a new property to store source references rather than using comments?

Lastly, you mentioned several technical points that are already on my to-do list 👍. I just need to find the time to progress with the work on this branch. And there is indeed some cleanup to be done in the martyrology (when I initially created it, I generated it from the calendars, but it still needs to be cleaned up).

**Reply by tukusejssirs** [reply]
> First, I noticed that we took different approaches to architecting Romcal. I used a functional codebase architecture, splitting the code into several packages based on functional scope, with shared components and configurations, and concentrating data in a specific package (@romcal/data). On the other hand, you propose the opposite: multiple packages per data type (calendars, locales, martyrology) but consolidating the Romcal codebase into a single package.

You say you _used a functional codebase architecture_. Well, I’ve thought that both are package lists are functional and both need to be consolidated (in a way) into a single library (not necesarily a package). As I see it, your `@romcal/data` package (or my `@romcal/martyrology` + `@romcal/*calendars` and `@romcal/locales`), as well as `@romcal/core`, are still just libraries and not apps, therefore they are still needed to be consolidated (i.e. built into) a single app. From this _consolidation_ point of view, we still build a single library _instance_ (via `new Romcal()` imported from `@romcal/core`) from multiple packages (using the data from `@romcal/data` (or from the three libraries I suggested above), while most of them are only libraries that can be imported by the main package (`@romcal/core`).

Regarding merging my proposed `@romcal/martyrology` + `@romcal/*calendars` and `@romcal/locales` into `@romcal/data`, I have no problem with that, although IMHO it would just be logical to use split packages (libraries) instead.

Okay, so my updated list of packages (libraries) is bellow.

- `@romcal/sources`:
  - it would contain all the martyrology/calendar/l10n sources (all data required for bibliographical reference to either a link a hard-copy publication) along with the reference (-s) to the original source (when available online;
  - ideally, all sources available online or to us should be copied to Google Drive and that link should be considered an archive);
  - no page numbers or chapter/section heading should be included;
  - potentially, this library could be merged into `@romcal/data`;
  - why I suggest creating a separate package for sources?
    - when done properly, there will be too many sources;
    - some sources will be cited multiple times;
    - we should simply reference an imported object and provide the page range, optionally with a chapter/section heading;
- `@romcal/data`:
  - it should contain:
    - [martyrology + event celebration metadata] all locale-agnostic celebration (meta-) data;
    - [celebrations] all celebration data to build the calendars in a particular GRC revision, both the 1969 GRC and all 1969 particular calendars;
    - [locales] everything needed to locale the l10n and calendar data, incl the default formatting per locale;
    - later (in the far future) we could also have `@romcal/grc1962/calendars` (cf #23) or `@romcal/grc1969-ordinariate-north-america/calendars` (cf #407);
  - it would reference the sources available in `@romcal/sources` with page number (-s), possibly also with a chapter/section heading when required for all martyrology, locales and celebrations data;
- `@romcal/core`:
  - it should use `@romcal/data` to build the requested calendar(s);
  - it should be calendar-agnostic, both revision-wise (`grc1962` vs `grc1969`) and rite-wise (Roman Catholic, Greek Catholic, Orthodox);
  - I understand that currently we support only GRC 1969 only, however, I still hope that in the far future, we could extend the support to other calendar types (revision-wise and rite-wise);
- `shared`:
  - it should contain all the utils, and shared data (like types) that are used in multiple packages.

And below is an updated diagram of how the data should flow from the data source to the utiliser package.

```mermaid
graph TD;
  S("@romcal/sources") --> D("@romcal/data");
  D --> R("@romcal/core");
```

> These two approaches can complement each other. In reality, the data is already organized as you presented but in various subfolders within the @romcal/data package.

Yeah, I know the data is already organised this way, but I’ve though it might be better to move the data out as separate libraries while we utilise a monorepo. IMHO moving each calendar revision/rite into a separate package/library from the beginning might simplify the creation of the calendars for different revisions/rites. I want to emphasise that the creation of such calendars is not aimed to be done in the near future, less so as an official package within `romcal` repository, however, IMHO the potential development should be taken into the design and architecture of the new monorepo structure (package-wise).

> In a few words, the goal of my architecture is to isolate each module and business layer of the Romcal library, making it easier to test each component, make them interchangeable and extendable by third-party components, and evolve them more independently. Thus, the aim is to use or extend Romcal's features and data more easily with third-party tools.

I fully agree with this. 

> So I've split consequently what is necessary for the proper functioning of the library, the data, and the validation and build tools into several packages.

What do you mean by the _proper versioning of the library_? In my opinion, we should use `semantic-release` and release a version of each publishable library (e.. `@romcal/core` and `@romcal/data`) with every commit/merge to the main branch (`dev`) regardless if there was any change in the code of a particular library in order to make sure the libraries are compatible (only libraries with same version should be supported, therefore, e.g. using `@romcal/core@3.0.0` and `@romcal/data@3.0.1` should not be officially supported, although the could work). `semantic-release` could automatically generate release notes, publish NPM packages, update `.version` in `package.json`, commit the changed files to the repository and create a Romcal release. IMHO this should be preferred to the current release system.

Moreover, we should also use `commitlint` to enforce the commit format.

We should also use `husky` to add some Git hooks (e.g. run `commitlint` before creating a commit).

> > as for `build`, it could be a separate package, I just have no idea how exactly you want to utilise it
>
> This package is used by almost all the other packages and contains common helpers for compiling the codebase or data of each package, generating the corresponding JSON schemas, and soon, it will enable us to publish properly each package independently on npmjs.

I think we should replace this approach with one that is more Nx-based.

In Nx, you usually have package commands which you can run as `nx $command $package`, e.g. `nx build core` would build `@romcal/core`. There are some pre-defined commands like `build`, `e2e`, `lint`, `serve`, `test`, but we could also create custom Nx commands.

I understand that building romcal is a nearly 300-line script, however, I think we should change this and make it buildable using a single line of code. I think it might be better to move to `@romcal/core` and export a function to build the romcal. That said, the easiest way is to simply run the script using an Nx command, but I don’t think it is the best way to solve this.

> > Add barrel `index.ts` files to re-export all the locales & calendars
>
> This is no longer necessary because the data is no longer in TS files but in YAML files, and it's imported differently by Romcal's build tools. If you want to access them, you should use the corresponding agnostics JSON files, which are built in the `@romcal/data/dist` directory.

Yeah, you want to convert the data to YAML (I prefer JSON, but more on that bellow, which can be re-exported in barrel files; I definitely agree with you to move the data from TS files though).

Anyway, why you want to store the data in YAML and then convert them to JSON during the build process? Wouldn’t it be easier to directly store the data in JSON? :thinking:

> > The martyrology is not related to the calendars
>
> Correction: In fact, Romcal's primary purpose is not to be a database but a library for generating liturgical calendars.

You’re right, I’ve forgotten about the fact that, e.g. a celebration could merge multiple persons from `martyrology.ts` into a single celebration.

> So the first-class citizen here is the calendar. From there, we derive the metadata of individuals/persons through the martyrology (which enriches the metadata of a celebration) and the locales (to translate all these metadata). The martyrology and locales are in service of the calendars, and not the other way around or independent entities.

Yeah, I get the data flow (I have always understood this, albeit I forgot some stuff of the implementation). I simply described the flow from data _source_ to data _utiliser_, you described in from data utiliser (calendar) to data sources (martyrology, locales). Usually, I prefer going through and algorithm from bottom (source) up (utiliser).

> Regarding support for other forms/rites, I've left several comments, including [this one](https://github.com/romcal/romcal/issues/23#issuecomment-1003461132). For now, I prefer to focus on modularizing the Romcal library, which will allow to easily "extend" Romcal to support such calendars but from a third-party repository/project.

Etienne, no need to re-iterate the fact that you don’t want to add the support for other revisions and rites. I get it. Also I don’t want to add it _rightaway_, however, I still think it would be a great addition _in the far future_, therefore we might take that into an account.

> Therefore, I am against the direct support of other form/rites within Romcal because the context, rules and logics are often very different. It is essential to handle and test them separately.

I get it, however, still it might be beneficial to move the 1969 GRC rules and logics to `@romcal/grc1969/calendars` (or `@romcal/data` if you wish) and keep `@romcal/core` revision- and rite-agnostic. However, it is not a hard requirement, just nice to have. 

> As for storing sources or source references, that's a great idea.

:pray:

> However, we need to consider copyright issues.

I didn’t mean to _store_ the source _copies_ into the repository. Yes, ideally, we should copy the sources to Google Drive (in can of copyright issues, only privately, without sharing it with the world) with a link to the original source in a text file (similarly how I do it already).

All I meant is to store the _metadata_ of the sources in `@romcal/sources` (or `@romcal/data`), like:

- source category/type (first/second/third party);
- URL;
- author;
- publication/book/artitle name;
- publishing date (exact date, year, century, etc);
- publishing place;
- (for online resources) date of access;
- volume.

This would be an object of `Source` type. Each source would get a unique ID.

Then, in calendar/locale/martyrology item data, we would use `Source` to create a _reference_ of `SourceReference` type by using the source ID and optionally adding some other metadata (page number, chapter/section title, etc). All this data could be used to create a bibliographical reference if wanted.

> For references, I believe they should be placed as close to the data as possible (closed to definitions of calendars/locales/martyrology...). Perhaps we should create a new property to store source references rather than using comments?

Indeed, I hate storing data in code comments.

I have imagined to create some types already (they could be improved):

<details>
  <summary>Example <code>Source</code> and <code>SourceReference</code> types, as well as their example values</summary>

```ts
type DateExact = {
  date: number,
  month: number,
  year: number
}

type DateYear = number

type DateType = DateExact | DateYear | {
  approximate?: boolean,
  type: 'century' | 'decade' | 'year',
  value: number
}

type Source = {
  accessDate?: DateExact,
  author?: string,
  date: DateType,
  edition?: number | string,
  id: string,
  language: string,
  name: string,
  note?: string,
  place?: string,
  type: 'first' | 'second' | 'third'
  url?: string,
  volume?: number
}

type Page = number | {
  from: number,
  to: number
}

/** @remarks Note that a chapter contains multiple sections. We might consider using `heading${n}` (e.g. `heading1`, `heading2` instead for easier understanding. */
type SourceReference = Source.id | {
  chapter?: string,
  note?: string,
  page?: Page[]
  section?: string,
  source: Source.id
}

// Examples of sources
// Note: This should be stored either in `@romcal/sources` or `@romcal/data`.
const sources: Source[] = [
  {
    id: 'mr_en_2011_ed3e_us',
    accessDate: {
      date: 15,
      month: 10,
      year: 2023
    },
    date: 2011,
    edition: `third typical edition, ammended`,
    language: 'English',
    name: 'The Roman Missal',
    note: 'For use in the dioceses of the United States of America',
    type: 'first'
  },
  {
    accessDate: {
      date: 15,
      month: 10,
      year: 2023
    },
    date: {
      date: 10,
      month: 10,
      year: 2023
    },
    id: wikipedia_general_roman_calendar,
    language: 'English',
    name: 'General Roman Calendar',
    url: 'https://en.wikipedia.org/w/index.php?title=General_Roman_Calendar&oldid=1179457480',
    type: 'third'
  }
]

// Examples of references
// Note: Each martyrology/calendar/locale data should be ideally accompanied with a reference like this.
const references: SourceReference[] = [
  // Valid for short articles or ones that deal with a single topic (e.g. a saint); it is genenerally invalid for printed/hard-copy sources (e.g. books, magazines, etc)
  'wikipedia_general_roman_calendar',
  // Prefered for longer articles
  {
    section: 'January',
    chapter: 'List of celebrations inscribed in the GRC',
    source: 'wikipedia_general_roman_calendar'
  },
	{
		section: 'January',
    page: 121,
    chapter: 'General Roman Calendar',
    source: 'mr_en_2011_ed3e_us'
  }
]
```

</details>

> Lastly, you mentioned several technical points that are already on my to-do list 👍. I just need to find the time to progress with the work on this branch. And there is indeed some cleanup to be done in the martyrology (when I initially created it, I generated it from the calendars, but it still needs to be cleaned up).

Well, the thing is that I wanted to start moving the data into an Nx monorepo, however, I wanted to get some consensus with you regarding some configuration. As I see it now, it might be better to do it without your approval and decide the details myself, then we can discuss the details in a PR and change what needs to be changed. Not my cup of tea, but I get it: we both a time-constrained. :wink:

Just a side question: if you don’t want to add metadata of event celebrations into `martyrology.ts`, where would you like to move them? I am againt deleting of such information (even though incomplete for now). We could use the following files (be it in JSON/TS/YAML files):
- `martyrology`/`persons`,
- `events`.

These files would be either in `@romcal/martyrology` or in a folder under `@romcal/data`.

> Streamline future data contributions: use a human-friendly data format (YAML).
>
> a human-friendly data format (YAML)

Regarding YAML vs JSON, I prefer using JSON. Yes, YAML has more features, however, unless we want to use YAML features that are not available in JSON (which IMHO we are not using, at least not yet), why do you prefer YAML?

IMHO, the following are JSON advantages why we should prefer it to YAML:

- it can be minified (unnecessary whitespace characters could be removed, e.g. during a ‘build’ step);
- `import info from './file.json' assert {type: 'json'};` does not require a loader or additional library;

Advantages of YAML:

- it has features that JSON does not have.

Both YAML and JSON are used widespread, therefore, it should be usable in any programming language without any issue.
